### ISSUE TITLE
Implement Russia Doll Cache

### ISSUE THREAD

**Issue by joaomdmoura** [rails-api/active_model_serializers#1045]
We have already implemented cache and fragment cache on 0.10.x, but we still need to make it better by implementing a Russia Doll Cache that will enable us to increase our performance when serializer different set of a resource.

**Comment by mib32** [comment]
Hey @joaomdmoura ! Thanks for all your work.
Could you elaborate, is this related to my desire to have separate cache sources, like in next gist.
The FullMarker inherits from marker, and it also inherits it's cache.
```
class FullMarkerSerializer < MarkerSerializer
  attributes :name, :special_offer
  belongs_to :category
  cache key: 'business_full_markers', only: [:name, :special_offer, :category]
  cache key: 'business_markers', except: [:name, :special_offer, :category]

  def special_offer
    object.sale.special_offer.try(:description)
  end

end
```

I'm correct that now it doesn't work?

**Comment by joaomdmoura** [comment]
No, @mib32 the Russian Doll cache is a technique of nesting fragment caches to maximize cache hits. By nesting fragment caches, it ensures that caches can be reused even when content changes. 
When a change occurs to the top-most fragment cache, only that cache must be expired. Every nested cache of the parent can be reused, which provides a significant performance increase. 

I'll need to check but what you posted might work now, not because of the inheritance, but because the cache is based on the model, and not on the serializer, anyway, I'm not so sure,I've to double check.

**Comment by passalini** [comment]
@joaomdmoura I'm trying to use this technic to improve the response time and the only way that I thought was caching the `as_json` from the `adapter`. This idea came up after I read [Hawkins](http://hawkins.io/2012/07/advanced_caching_part_6-fast_json_apis/)'s posts and changed that a little bit.

Right know I have something like this:
```ruby
class ActiveModel::Serializer::Adapter::FlattenJsonWithCustomCache < ActiveModel::Serializer::Adapter::FlattenJson
  def as_json(options = nil)
    Rails.cache.fetch(serializer.object, tag: cache_tags) do
      super
    end
  end

  private
    def cache_tags
      [get_object_superclass.to_s.underscore, serializer.object.id].join('_')
    end

    def get_object_superclass(klass = serializer.object.class)
      return klass if klass.superclass == ActiveRecord::Base
      get_object_superclass(klass.superclass)
    end
end
```

```ruby
# initializers/ams.rb
ActiveModel::Serializer.config.adapter = ActiveModel::Serializer::Adapter::FlattenJsonWithCustomCache
```

ps: I added `cashier` to expire that cache by tag because I have many relations by object and put a `touch` at all relations isn't a good idea for my case.

I hope that it helps with something and if someone has ideas to improve this workaround I will be very happy! =D

**Comment by passalini** [comment]
@joaomdmoura I had to change the cache to the `serializable_hash` method. I did it to use russian doll  with collection objects, with it we can cache (russian doll) object by object... We can do some code to russian doll all objects in one cache, is just use the code above with my new approach

**Comment by bf4** [comment]
ref: https://github.com/rails-api/active_model_serializers/pull/1372
